File: app/src/main/java/com/donald/musictheoryapp/Music/MusicXML/Key.java
package com.donald.musictheoryapp.Music.MusicXML;

import org.json.JSONException;
import org.json.JSONObject;

public class Key
{
	public static class Mode
	{
		public static final int MINOR = 0;
		public static final int MAJOR = 1;

		public static final String[] STRINGS =
			{
				"minor",
				"major"
			};

		public static int fromString(String modeString)
		{
			for(int i = 0; i < STRINGS.length; i++)
			{
				if(STRINGS[i].equals(modeString)) return i;
			}
			return -1;
		}
	}

	public final int fifths;
	public final int mode;

	public Key(int fifths, int mode)
	{
		this.fifths = fifths;
		this.mode = mode;
	}

	public static Key fromJSON(JSONObject key) throws JSONException
	{
		return new Key(
			key.getInt("fifths"),
			Mode.fromString(key.getString("mode"))
		);
	}
}

File: app/src/main/java/com/donald/musictheoryapp/music/musicxml/Key.kt
package com.donald.musictheoryapp.music.musicxml

import org.dom4j.Element
import kotlin.Throws
import org.json.JSONException
import org.json.JSONObject
import java.lang.IllegalArgumentException

class Key(val fifths: Int, val mode: Int) {

    object Mode {
        const val MINOR = 0
        const val MAJOR = 1
        private val STRINGS = arrayOf(
            "minor",
            "major"
        )

        fun stringOf(mode: Int): String {
            if (mode < MINOR || mode > MAJOR) throw IllegalArgumentException()
            return STRINGS[mode]
        }

        @JvmStatic
		fun fromString(modeString: String): Int {
            for (i in STRINGS.indices) {
                if (STRINGS[i] == modeString) return i
            }
            return -1
        }
    }

    fun addToXml(attributes: Element) {
        attributes.addElement("key").apply {
            addElement("fifths").addText(fifths.toString())
            addElement("mode").addText(Mode.stringOf(mode))
        }
    }

    fun clone(): Key {
        return Key(fifths, mode)
    }

    companion object {
        @Throws(JSONException::class)
        fun fromJson(key: JSONObject): Key {
            return Key(
                key.getInt("fifths"),
                Mode.fromString(key.getString("mode"))
            )
        }
    }

}